diff --git a/lib/parser.js b/lib/parser.js
index d72671a20b0002dcb0da03dcc50712a55c2572e0..358ae614be22b3dcbfce538dcedb2e5e48c96629 100644
--- a/lib/parser.js
+++ b/lib/parser.js
@@ -134,7 +134,7 @@ function RE(s, flag) {
 var matchIdentifierAtIndex;
 if (REGEX_SUPPORTS_U_AND_Y) {
     // Native
-    var IDENTIFIER_PREFIX_RE_1 = RE('([^\\p{White_Space}\\p{Pattern_Syntax}]*)', 'yu');
+    var IDENTIFIER_PREFIX_RE_1 = RE('((?:[^\\p{White_Space}\\p{Pattern_Syntax}]|\\.)*)', 'yu');
     matchIdentifierAtIndex = function matchIdentifierAtIndex(s, index) {
         var _a;
         IDENTIFIER_PREFIX_RE_1.lastIndex = index;
@@ -646,6 +646,22 @@ var Parser = /** @class */ (function () {
                     };
                 }
             }
+            case 'tpl': {
+              var argCloseResult = this.tryParseArgumentClose(openingBracePosition);
+              if (argCloseResult.err) {
+                  return argCloseResult;
+              }
+              var location_1 = createLocation(openingBracePosition, this.clonePosition());
+
+              return {
+                val: {
+                    type: TYPE.template,
+                    value: value,
+                    location: location_1,
+                },
+                err: null,
+              };
+            }
             default:
                 return this.error(ErrorKind.INVALID_ARGUMENT_TYPE, createLocation(typeStartPosition, typeEndPosition));
         }
diff --git a/lib/types.d.ts b/lib/types.d.ts
index f7e4cb84dfba2c4c1871e7518ed7f47db0761be3..5bbb658521f6b33b4fbbac8cd896d92a4c0ffaad 100644
--- a/lib/types.d.ts
+++ b/lib/types.d.ts
@@ -40,7 +40,11 @@ export declare enum TYPE {
     /**
      * XML-like tag
      */
-    tag = 8
+    tag = 8,
+    /**
+     * Include another i18n entry given by its key
+     */
+    template = 9
 }
 export declare enum SKELETON_TYPE {
     number = 0,
@@ -74,6 +78,7 @@ export interface SimpleFormatElement<T extends TYPE, S extends Skeleton> extends
 export type NumberElement = SimpleFormatElement<TYPE.number, NumberSkeleton>;
 export type DateElement = SimpleFormatElement<TYPE.date, DateTimeSkeleton>;
 export type TimeElement = SimpleFormatElement<TYPE.time, DateTimeSkeleton>;
+export type TemplateElement = BaseElement<TYPE.template>;
 export type ValidPluralRule = 'zero' | 'one' | 'two' | 'few' | 'many' | 'other' | string;
 export interface PluralOrSelectOption {
     value: MessageFormatElement[];
@@ -91,7 +96,7 @@ export interface PoundElement {
     type: TYPE.pound;
     location?: Location;
 }
-export type MessageFormatElement = ArgumentElement | DateElement | LiteralElement | NumberElement | PluralElement | PoundElement | SelectElement | TagElement | TimeElement;
+export type MessageFormatElement = ArgumentElement | DateElement | LiteralElement | NumberElement | PluralElement | PoundElement | SelectElement | TagElement | TimeElement | TemplateElement;
 export interface NumberSkeleton {
     type: SKELETON_TYPE.number;
     tokens: NumberSkeletonToken[];
@@ -117,6 +122,7 @@ export declare function isSelectElement(el: MessageFormatElement): el is SelectE
 export declare function isPluralElement(el: MessageFormatElement): el is PluralElement;
 export declare function isPoundElement(el: MessageFormatElement): el is PoundElement;
 export declare function isTagElement(el: MessageFormatElement): el is TagElement;
+export declare function isTemplateElement(el: MessageFormatElement): el is TemplateElement;
 export declare function isNumberSkeleton(el: NumberElement['style'] | Skeleton): el is NumberSkeleton;
 export declare function isDateTimeSkeleton(el?: DateElement['style'] | TimeElement['style'] | Skeleton): el is DateTimeSkeleton;
 export declare function createLiteralElement(value: string): LiteralElement;
diff --git a/lib/types.js b/lib/types.js
index 302f1cccc809bd46d418d0673bf8c102241aa349..90994d9c1d97daa620fb14162a3b0fd4e7dfa9ce 100644
--- a/lib/types.js
+++ b/lib/types.js
@@ -37,6 +37,10 @@ export var TYPE;
      * XML-like tag
      */
     TYPE[TYPE["tag"] = 8] = "tag";
+    /**
+     * Include another i18n entry given by its key
+     */
+    TYPE[TYPE["template"] = 9] = "template";
 })(TYPE || (TYPE = {}));
 export var SKELETON_TYPE;
 (function (SKELETON_TYPE) {
@@ -73,6 +77,9 @@ export function isPoundElement(el) {
 export function isTagElement(el) {
     return el.type === TYPE.tag;
 }
+export function isTemplateElement(el) {
+    return el.type === TYPE.template;
+}
 export function isNumberSkeleton(el) {
     return !!(el && typeof el === 'object' && el.type === SKELETON_TYPE.number);
 }
diff --git a/parser.js b/parser.js
index c50ac756acd6630ab19bcb7a3c2a33742afd7913..43af03834e324c649f9b256864b1a01074847438 100644
--- a/parser.js
+++ b/parser.js
@@ -137,7 +137,7 @@ function RE(s, flag) {
 var matchIdentifierAtIndex;
 if (REGEX_SUPPORTS_U_AND_Y) {
     // Native
-    var IDENTIFIER_PREFIX_RE_1 = RE('([^\\p{White_Space}\\p{Pattern_Syntax}]*)', 'yu');
+    var IDENTIFIER_PREFIX_RE_1 = RE('((?:[^\\p{White_Space}\\p{Pattern_Syntax}]|\\.)*)', 'yu');
     matchIdentifierAtIndex = function matchIdentifierAtIndex(s, index) {
         var _a;
         IDENTIFIER_PREFIX_RE_1.lastIndex = index;
@@ -649,6 +649,22 @@ var Parser = /** @class */ (function () {
                     };
                 }
             }
+            case 'tpl': {
+              var argCloseResult = this.tryParseArgumentClose(openingBracePosition);
+              if (argCloseResult.err) {
+                  return argCloseResult;
+              }
+              var location_1 = createLocation(openingBracePosition, this.clonePosition());
+
+              return {
+                val: {
+                    type: types_1.TYPE.template,
+                    value: value,
+                    location: location_1,
+                },
+                err: null,
+              };
+            }
             default:
                 return this.error(error_1.ErrorKind.INVALID_ARGUMENT_TYPE, createLocation(typeStartPosition, typeEndPosition));
         }
diff --git a/types.d.ts b/types.d.ts
index f7e4cb84dfba2c4c1871e7518ed7f47db0761be3..5bbb658521f6b33b4fbbac8cd896d92a4c0ffaad 100644
--- a/types.d.ts
+++ b/types.d.ts
@@ -40,7 +40,11 @@ export declare enum TYPE {
     /**
      * XML-like tag
      */
-    tag = 8
+    tag = 8,
+    /**
+     * Include another i18n entry given by its key
+     */
+    template = 9
 }
 export declare enum SKELETON_TYPE {
     number = 0,
@@ -74,6 +78,7 @@ export interface SimpleFormatElement<T extends TYPE, S extends Skeleton> extends
 export type NumberElement = SimpleFormatElement<TYPE.number, NumberSkeleton>;
 export type DateElement = SimpleFormatElement<TYPE.date, DateTimeSkeleton>;
 export type TimeElement = SimpleFormatElement<TYPE.time, DateTimeSkeleton>;
+export type TemplateElement = BaseElement<TYPE.template>;
 export type ValidPluralRule = 'zero' | 'one' | 'two' | 'few' | 'many' | 'other' | string;
 export interface PluralOrSelectOption {
     value: MessageFormatElement[];
@@ -91,7 +96,7 @@ export interface PoundElement {
     type: TYPE.pound;
     location?: Location;
 }
-export type MessageFormatElement = ArgumentElement | DateElement | LiteralElement | NumberElement | PluralElement | PoundElement | SelectElement | TagElement | TimeElement;
+export type MessageFormatElement = ArgumentElement | DateElement | LiteralElement | NumberElement | PluralElement | PoundElement | SelectElement | TagElement | TimeElement | TemplateElement;
 export interface NumberSkeleton {
     type: SKELETON_TYPE.number;
     tokens: NumberSkeletonToken[];
@@ -117,6 +122,7 @@ export declare function isSelectElement(el: MessageFormatElement): el is SelectE
 export declare function isPluralElement(el: MessageFormatElement): el is PluralElement;
 export declare function isPoundElement(el: MessageFormatElement): el is PoundElement;
 export declare function isTagElement(el: MessageFormatElement): el is TagElement;
+export declare function isTemplateElement(el: MessageFormatElement): el is TemplateElement;
 export declare function isNumberSkeleton(el: NumberElement['style'] | Skeleton): el is NumberSkeleton;
 export declare function isDateTimeSkeleton(el?: DateElement['style'] | TimeElement['style'] | Skeleton): el is DateTimeSkeleton;
 export declare function createLiteralElement(value: string): LiteralElement;
diff --git a/types.js b/types.js
index e638e6ee7c147735fab127755300d6c03742efb0..4441106e4fff884ff5230f6dad19104574531763 100644
--- a/types.js
+++ b/types.js
@@ -40,6 +40,10 @@ var TYPE;
      * XML-like tag
      */
     TYPE[TYPE["tag"] = 8] = "tag";
+    /**
+     * Include another i18n entry given by its key
+     */
+    TYPE[TYPE["template"] = 9] = "template";
 })(TYPE = exports.TYPE || (exports.TYPE = {}));
 var SKELETON_TYPE;
 (function (SKELETON_TYPE) {
@@ -85,6 +89,10 @@ function isTagElement(el) {
     return el.type === TYPE.tag;
 }
 exports.isTagElement = isTagElement;
+function isTemplateElement(el) {
+    return el.type === TYPE.template;
+}
+exports.isTemplateElement = isTemplateElement;
 function isNumberSkeleton(el) {
     return !!(el && typeof el === 'object' && el.type === SKELETON_TYPE.number);
 }